### PR TITLE
UPSTREAM: 58293: fail earlier on discovery failures

### PR DESCRIPTION
**Pull request by deads2k**
Picking a fail-early pull for when we're unable to lookup resource information from the apiserver.

/assign juanvallejo

### CODE REVIEW

**Comment by deads2k** [comment]
@juanvallejo this is exposing a problem in `oc process` and discovery mapper lookups.  Can you chase it down?

**Comment by juanvallejo** [comment]
@deads2k sure, taking a look

File: vendor/k8s.io/kubernetes/pkg/kubectl/cmd/util/factory_object_mapping.go [hunk 1]
```diff
@@ -77,21 +77,25 @@ func NewObjectMappingFactory(clientAccessFactory ClientAccessFactory) ObjectMapp
...
 	if err != nil && !discoveryClient.Fresh() {
 		discoveryClient.Invalidate()
 		groupResources, err = discovery.GetAPIGroupResources(discoveryClient)
 	}
+	// even if we can't get all the results, we're better off continuing with what we can than not presenting
+	// anything.  This mirrors old behavior that used to fallback to a hardcoded list of API types compiled
+	// into kubernetes.
 	if err != nil {
-		glog.V(3).Infof("Unable to retrieve API resources, falling back to hardcoded types: %v", err)
-		return legacyscheme.Registry.RESTMapper(), legacyscheme.Scheme, nil
```
Comment (juanvallejo) [inline vendor/k8s.io/kubernetes/pkg/kubectl/cmd/util/factory_object_mapping.go]: Error is happening here: when we run `oc process` expecting not to contact the server, returning a mapper for hardcoded types would suffice. With this patch, we are actually continuing on, and using a deferred discovery mapper down below.

**Comment by deads2k** [comment]
@juanvallejo  see the second commit.

**Comment by juanvallejo** [comment]
@deads2k opened PR upstream: https://github.com/kubernetes/kubernetes/pull/59118

Some tests in /test cmd are failing with panics, looking..

Locally, `oc label` panics as well.

**Pull request closed without merging**